### ISSUE TITLE
SNOW-669163: Missing connection_invalidated / is_disconnect detection

### ISSUE THREAD

**Issue by kotofos** [snowflakedb/snowflake-sqlalchemy#348]
When DBAPI error raised it has `connection_invalidated` flag which can be used to detect disconnects and do the reconnect. This flag is dialect-dependent.
In core supported sqlalchemy dialects, like MySQL this feature works as expected.

## What is the current behavior?
Currently snowflake-sqlalchemy does not set this flag. It is hardcoded to False.

See code:
Default implementation: https://github.com/sqlalchemy/sqlalchemy/blob/86e21c2d6f750f55c5dbe31d6ab8875230d8ecb0/lib/sqlalchemy/engine/default.py#L749
Mysql implementation: https://github.com/sqlalchemy/sqlalchemy/blob/86e21c2d6f750f55c5dbe31d6ab8875230d8ecb0/lib/sqlalchemy/dialects/mysql/base.py#L2612

Snowflake dialect is missing it and just uses default inherited behavior.

## What is the desired behavior?
 `connection_invalidated` is set when there was connection problems.

## How would this improve `snowflake-connector-python`?
This will allow better disconnect detection and allow generic reconnect code to work.

Something like
```
            try:
                session.execute('stmt')
            except DBAPIError as e:
                if e.connection_invalidated:
                    # retry

```
which works for other dialects.

## References, Other Background
Found cases when this is happening
- Auth token expired after 4 hours of inactivity. `(snowflake.connector.errors.ProgrammingError) 390114 (08001): None: Authentication token has expired.  The user must authenticate again.`
- Session expired after configured idle time. `snowflake.connector.errors.ProgrammingError: 390111: Session no longer exists.  New login required to access the service.`
- session killed from server side
- user deleted/disabled and then reenabled

**Comment by sfc-gh-sfan** [comment]
Makes total sense to me. We should probably mimic the mysql behavior, but we also need to make sure we handle it correctly in snowflake's context.

**Comment by sfc-gh-mraba** [comment]
Resolved by #740 (merged in c9d7c6b). The dialect's `is_disconnect` now recognizes expired-session/token and closed-connection error codes, so SQLAlchemy sets `connection_invalidated` and recycles pooled connections instead of reusing a dead one.
